### ISSUE TITLE
Context Manager that disables training mode with in a nn.Module.

### ISSUE THREAD

**Issue by WillBrennan** [pytorch/pytorch#18220]
## 🚀 Context Manager that disables training mode with in a nn.Module.
Currently `with torch.no_grad():` can be used to disable gradient calculation temporarily while batchnorm and drop out layers have to be set to evaluation mode using `model.train(True)` followed by a `model.train(False)` once evaluation is complete. Currently this syntax is used, 

```python
model.train(False)
with torch.no_grad():
  outputs = model(inputs)
  model.train(True)
  return outputs
```
. This is a potential source of bugs, as people will sometimes instead use `model.eval()` and forget to set the model back to training mode at the end of inference. Suggest having the following syntax, 

```python
with torch.no_train(model):
  with torch.no_grad():
    outputs = model(inputs):
    return outputs
```
. This syntax would stop this type of bug from occurring.

**Comment by ezyang** [comment]
This would be pretty easy to implement, but we're not sure how widely used it would be. To others: if this is something that you would find useful, please thumbs up on this issue.

A trivial implementation of `no_train` is:

```
class no_train(object):
  def __init__(self, module):
    self.module = module
  def __enter__(self):
    self.module.train(false)
  def __exit__(self):
    self.module.train(true)
```

Though, I guess a real implementation has to be a bit more careful to restore the *original* training state, rather than just setting it to `true`

**Comment by shapovalov** [comment]
Another implementation: https://discuss.pytorch.org/t/opinion-eval-should-be-a-context-manager/18998/3

**Comment by TamirCohen** [comment]
Is this feature wanted? If so I can open a PR

**Comment by WillBrennan** [comment]
I'd personally find it very helpful!

**Comment by albanD** [comment]
From some discussion on this, the request here (that takes a single Module as input) is pretty easy to do locally if you need it (see comment from ezyang above) and has limited use.
In particular, since in this context you don't want to assume the original state of the Module in either your train or eval code. You need to use this context manager in both of them. But if you're doing that, then setting train/eval mode there will have the same behavior.

During the discussion, it was mentioned that a context manager that would affect all Modules within the scope (and not just the one passed in) would be more useful but at first sight it seems like the technical complexity of such approach will be prohibitive to actually do it.

So I don't think we'll be moving forward with this at the moment and would recommend you use the example from https://github.com/pytorch/pytorch/issues/18220#issuecomment-476303873 in your local code if you need this.
